CSS Style Changes and Note Function - SafeMemo for ESP3D-WEBUI

jayjojayson opened the pull request:
I made some Changes in the css of the index.html.gz to get a better design.

luc-github commented:
yes UI is nice
but changing index.html.gz is actually useless as it is result of build process, it will be rebuilded at next change.
1 -  the way to go is to modify source css 
2 -  because your UI, even indeed nice, is pretty different , need to provide a way to keep existing UI , theme is not supported  yet, so need to change css at compile time to select which one to use, I would suggest a parameter in gulp script to select the target css and so the UI  - and add it to build batch so all UI will be created  at once

have a look here : https://github.com/luc-github/ESP3D-WEBUI/wiki/Compilation---Development
here : https://github.com/luc-github/ESP3D-WEBUI/blob/2.1/buildLanguagePacks.bat
and here : https://github.com/luc-github/ESP3D-WEBUI/blob/2.1/gulpfile.js

luc-github commented:
also all the panel title are white on grey which is little bit too light IMHO
![image](https://user-images.githubusercontent.com/8822552/73121968-0810b100-3fba-11ea-88f9-e18ee2ee5a5e.png)

jayjojayson commented:
Thanks and thanks for your quick reply. I have done this for my own and thought to share it is always good. I know, that these changes are not rebuilded at next change, but didn't know where put these changes, until now.  I don't have much knowledge about programming except html and some java-basics, but i will look into your links and see what i can do. I will give you a feedback, after studing it.

The panel title "white on grey" can be changed quickly. At this time all headers (h1...h6) have the same color. 

Thanks for your lovely work on this project!

jayjojayson commented:
Can i use the the index.html.gz in languages/multi to make changes on html himself, not on css? I think out of this file the other language-index-files are build up, or i'm wrong?

luc-github commented:
sources are here : https://github.com/luc-github/ESP3D-WEBUI/tree/2.1/www
check css subdir

gulp script build html according parameters
bat file build all index.htm.gz using gulp script with different parameters for language

jayjojayson commented:
I have changed the css files according to my design and the pratice of github is much more clearer to me now. Will see, what i can do more for this project. :)

luc-github commented:
I also suggest to test on different pc browser as well as phone and tablet because resolution has impact on UI

jayjojayson commented:
I had no Problems with that bigger Temperaturpanel, test it on Chrome, Phone an Tablet. But if i made the panel even bigger, as now, i got problems with the resolution

jayjojayson commented:
Hi Luc, first of all i have added a seccond way to the wiki page to connect the wemos d1 mini with logic level converter : https://github.com/luc-github/ESP3D/wiki/D1-mini

Added macro-examples in my copy of this project, for Laser ON/OFF with Gcode M106 and M107

Some time ago i have done a little project to save notes in localStorage. I thought this would be a nice addtion for this project. So i have included my project in to yours and adapted it for the current WebUI. I hope, i have done everything correctly. There are some new Files memotab.html, memo.css, memo.js and one png. The other pictures are done with svg. I have edited the index.html, tab_bar.html and preferencesdlg.html to integrate the memotab.html. In the preferencesdlg.js i don't know exactly what i have to change, so i didn't do anything there. But i know the Checkbutton for memo doesn't work at the moment. Maybe you can fix this.

<a href="https://ibb.co/pJtZ5PB" target="_blank"><img src="https://i.ibb.co/SRh0Gx4/Unbenannt.png" alt="Unbenannt" border="0"></a>

This is a simple Note-Function for the WebUI, but it can make new notes, can edit existing notes, delete individual note or delete all notes. Every note has a subject and note. So if you put the subject of any note in the subject-field and click load, then you get the associated Note from local Storage. All notes are shown on the memotab.html.

jayjojayson commented:
Now if i want test the new WebUI, i have to do a full flash of the sketch from ESP3D Project or is there a way to update / refresh the WebUI without flashing the wemos d1 mini. I don't think so, but maybe there is a way.

luc-github commented:
what the purpose of the SafeMemo in webUI ?

jayjojayson commented:
The purpose is to make notes / memos inside the webui. I thought this is a good idea. So everyone can make notes for example to save some gcode examples or something else. The SafeMemo tool looks like in the picture above.

luc-github commented:
Hmm not sure to understand the user case
also localstorage means you only use one browser and only one computer
so if you use phone then PC you loose the data set on phone, that is why I moved UI settings from localstorage to SPIFFS so user keep data from where ever   he connects
also localstorage has some restrictions like not working on some captive portal OS, I forget if it crash or is silently ignored

jayjojayson commented:
Yes you are right, the localStorage works only on one PC and loose the data if some other device is used. So your suggestion makes senses, but i don't know how to save the data in SPIFFS. I'm really happy that i have did it so far. Maybe you can fix it in the memo.js to save the data in SPIFFS if you have some free time.
For me and my test it will be enough, because i'm using only one PC. I have to follow these stepps to generate the new index.html.gz, right? I will test it in the evening.
[https://github.com/luc-github/ESP3D-WEBUI/wiki/Compilation---Development](https://github.com/luc-github/ESP3D-WEBUI/wiki/Compilation---Development)

luc-github commented:
yes need to build it to test it

jayjojayson commented:
Yeah, i have managed to build up the index.html. The first Test was not good, i had to change the code it a little bit to get a right looking html page, but it is working fine now. :) Some more CSS changes was made, too. See my pictures, maybe you have now a clearer view, what i mean with notes. I find often code/gcode in web and want to save it and to do it in the WebUI is great plus for me.

I would be glad if you implemented it in the next ESP3D-WebUI version. So I contributed a small part to this project.

<a href="https://ibb.co/Nm5tQpX"><img src="https://i.ibb.co/7RPj8VH/Unbenannt1.png" alt="Unbenannt1" border="0"></a>
<a href="https://ibb.co/vB9KqCn"><img src="https://i.ibb.co/KGn3hcv/Unbenannt2.png" alt="Unbenannt2" border="0"></a>
<a href="https://ibb.co/9wcpKT2"><img src="https://i.ibb.co/Q8pJSNn/Unbenannt3.png" alt="Unbenannt3" border="0"></a>
<a href="https://ibb.co/mCmHqJM"><img src="https://i.ibb.co/ccPNFJ7/Unbenannt4.png" alt="Unbenannt4" border="0"></a>
<a href="https://ibb.co/ky9kjbc"><img src="https://i.ibb.co/qnWXKq1/Unbenannt5.png" alt="Unbenannt5" border="0"></a>
<a href="https://ibb.co/LhRnps9"><img src="https://i.ibb.co/DgW5Kjw/Unbenannt6.png" alt="Unbenannt6" border="0"></a>

jayjojayson commented:
Hi, what you think about my styling and note-function?

luc-github commented:
I did not try yet as I am traveling, it is another style, about memo function I am not really convinced sorry, as it will be an alternate UI it may not be a problem
I will see how to modify scripts to keep existing and add this UI, do you have a name for this UI ?

jayjojayson commented:
No stress, have a nice travel. I have no problem, if you don't want to use the safeMemo. For me is the Memo a nice have, but i'm happy if my design is comming to your version. What you think about Iweb, weBox or boxUi as name. Currently i'm working on the top navigation. If i'm finish, i show the picture with the result.

jayjojayson commented:
hi luc, i have made some changes again, to fix some issues. Now i have tried to implement a clock in the header. Therefore i added a new js file called clock.js. Now after uploading it works, but i'm stuck on the loading page. Is there any Chance to upload the index.html.gz without the webui? Or can i disable the clock.js script via Chrome Dev Tools? 

<a href="https://ibb.co/jTK2jBN"><img src="https://i.ibb.co/VtyZkr1/Unbenannt.jpg" alt="Unbenannt" border="0"></a>

In the DevTools i can see, that the script is working. The Clock is shown on the loading page. But i can't find the script on the page to deactivate it and move on. Any help would be nice, but i think i have to flash the whole sketch again.

luc-github commented:
you can send [ESP710]FORMAT by serial or telnet to format SPIFFS and so remove the not working index.html.gz and then you will go to fallback page

or you can directly force the fallback page by typing 
`<your ip>?forcefallback=yes`

the fallback page has file uploader and FW updater

jayjojayson commented:
Thanks, that you have build in the forcefallback! :) Works great, now i can continue testing

jayjojayson commented:
Hi Luc, i made a lot of changes on the weekend. First i have split the Header and the Tabs/Navbar. The Menu tabs are now on the bottom. At 1 Colum the Menu is hidden and is going to the Header (hamburger menu). 
https://raw.githubusercontent.com/jayjojayson/ESP3D-WEBUI/2.1/images/1-colum.JPG

At 2,3 and 4 Colum the Menutabs are fixed to the bottom. 
https://raw.githubusercontent.com/jayjojayson/ESP3D-WEBUI/2.1/images/2-colum-better.JPG

The Pringprogress-Bar is imported from the other request (Added SD printing progress bar to the files panel) and works with Command M27 instead of M27 C. Now only the bytes are sending, but it's enough to show the progress. The current File name would be fine, but is not a must have for me.

Changes:
- added custom menu on 1 and 2,3,4 Colum
- 1 colum = hamburger menu
- 2,3,4 Colum Menu fixed to bottom
- fixed the Macro buttons-width in controls-panel at all media width
- added printprogress-bar
- added clock to the header
- updated language tpl and de
- Style Changes and some Design Changes
- Changed some svg

Please give it a try and have a look here, there you can find all the example images
https://github.com/jayjojayson/ESP3D-WEBUI/blob/2.1/README.md

Right now, the only problem i have, is to fix the hamburger- and bottom-menu content. I have hide the bottom menu with css at specific width and added a copy of the menu (slightly adjusted) to the navbar. Now there are two menus with same content and the option "hide it" or not on the produced index.html.gz. Therefore the tabs.js is not working right at the moment, because they are two menupoints with the same link, like described. The hamburger menu is showing all menutabs, but in the bottommenu are only the tabs with no "hide it" displayed. But i don't know what to change/adjust in tabs.js to make it working. Please can you have a look, if you have time and tell me what i have to adjust? Thanks a lot.

Now i think, the Menudesign looks clean and modern and it is small, to see more content on first view. Don't know how to say it better. Tested it with Chrome win10, Safari on Ipad mini 2 and Chrome on Android Phone.

luc-github commented:
About progress bar: same question as in previous PR
did you check with smoothieware and repetier ?
must be hidden for GRBL/GRBL_ESP32

About tabs:
I see your menu show all even camera which is an option in preferences - do you display according preferences?
I am not sure what is your problem but I guess when using tabs.js it update current selected tabs so  your menu is not a tabs - you may need to update tabs.js to handle command from menu in addition of a command from tabs, another quick and dirty solution is to ask your menu item to 'click' the corresponding tabs

jayjojayson commented:
sorry..  jayjojayson requested a review from luc-github 1 minute ago
i have clicked the wrong button.. sometimes these menus on github are confusing for me.. ;)

printprogress:
no, i don't test it with smoothieware and repetier. I have only my anycubic with marlin firmware. But if they use the same Command M27 to get the current printing byte, it should work, too.

Tabs:
yes i display the according preference. So the Menu at 1 colum is the correct menu, that should also displayed at 2,3 and 4 colum. I think you are right with that "you may need to update tabs.js to handle command from menu in addition of a command from tabs, another quick and dirty solution is to ask your menu item to 'click' the corresponding tabs". Will see, what i can do, but my javascript knowledge is not very much.

luc-github commented:
if you do not test all FW target I suggest to limit the feature to  the FW you test
as wrote must be hidden for GRBL/GRBL_ESP32, as GRBL does not support this feature at all

jayjojayson commented:
In my readme i refer several times to having only tested it with marlin firmware and that the progressbar works with G-Command M27. In my features-list it is listed also with (Marlin Firmware).

luc-github commented:
per my experience very few people read the readme, they will just complain it is not working as they expect as WebUI is for all targeted FW as it is stated in your read me : https://github.com/jayjojayson/ESP3D-WEBUI#what-is-that-

jayjojayson commented:
ok, i know what you mean. I have delete the Part "What is that" for no confusion.
